netadapter/netadapter.go

```diff
@@ -43,31 +61,118 @@ func (na *NetAdapter) Start() error {
 
 // Stop safely closes the NetAdapter
 func (na *NetAdapter) Stop() error {
+	if atomic.AddInt32(&na.stop, 1) != 1 {
+		log.Warnf("Net adapter stopped more than once")
+		return nil
+	}
 	return na.server.Stop()
 }
 
-func (na *NetAdapter) newPeerConnectedHandler() server.PeerConnectedHandler {
-	return func(connection server.Connection) {
-		peer := NewPeer(connection)
-		router, err := na.routerInitializer(peer)
+func (na *NetAdapter) newOnConnectedHandler() server.OnConnectedHandler {
+	return func(connection server.Connection) error {
+		router, err := na.routerInitializer()
 		if err != nil {
-			// TODO(libp2p): properly handle error
-			panic(err)
+			return err
+		}
+		connection.SetOnDisconnectedHandler(func() error {
+			na.unregisterConnection(connection)
+			return router.Close()
+		})
+		router.SetOnIDReceivedHandler(func(id *ID) {
+			na.registerConnection(connection, router, id)
+		})
+
+		na.startReceiveLoop(connection, router)
+		na.startSendLoop(connection, router)
+		return nil
+	}
+}
+
+func (na *NetAdapter) registerConnection(connection server.Connection, router *Router, id *ID) {
+	na.connectionIDs[connection] = id
+	na.idsToConnections[id] = connection
+	na.idsToRouters[id] = router
+}
+
+func (na *NetAdapter) unregisterConnection(connection server.Connection) {
+	id, ok := na.connectionIDs[connection]
+	if !ok {
+		return
+	}
+
+	delete(na.connectionIDs, connection)
+	delete(na.idsToConnections, id)
+	delete(na.idsToRouters, id)
+}
+
+func (na *NetAdapter) startReceiveLoop(connection server.Connection, router *Router) {
+	spawn(func() {
+		for {
+			if atomic.LoadInt32(&na.stop) != 0 {
+				err := connection.Disconnect()
+				if err != nil {
+					log.Warnf("Failed to disconnect from %s: %s", connection, err)
+				}
+				return
+			}
+
+			message, err := connection.Receive()
+			if err != nil {
+				log.Warnf("Failed to receive from %s: %s", connection, err)
+				err := connection.Disconnect()
```

Comment:
Since you are disconnecting, you should probably return afterwards, otherwise the loop is stuck forever